proj4/man/man1/proj.1
.\" release 4
.nr LL 5.5i
.ad b
.hy 1
.TH PROJ 1 "2000/03/21 Rel. 4.4"
.SH NAME
proj \- forward cartographic projection filter
.br
invproj \- inverse cartographic projection filter
.SH SYNOPSIS
.B proj
[
.B \-bcCeEfiIlmorsStTvVwW
[
.I args
] ] [
.B +args
]
file[s]
.br
.B invproj
[
.B \-bcCeEfiIlmorsStTwW
[
.I args
] ] [
.B +args
]
file[s]
.SH DESCRIPTION
.I Proj
and
.I invproj
perform respective forward and inverse transformation of cartographic data
to or from cartesian data with a wide range of selectable projection functions.
.PP
The following control parameters can appear in any order:
.TP
.BI \-b
Special option for binary coordinate data input and output
through standard input and standard output.
Data is assumed to be in system type
.I double
floating point words.
This option is to be used when
.B proj
is a
.I son
process and allows bypassing formatting operations.
.TP
.BI \-i
Selects binary input only (see
.B \-b option).
.TP
.BI \-C
Check. Invoke all built in self tests and report.
Get more verbose report by preceding with the
.B \-V option).
.TP
.BI \-I
alternate method to specify inverse projection.
Redundant when used with
.B invproj.
.TP
.BI \-o
Selects binary output only (see
.B \-b option).
.TP
.BI \-t "a"
.I A
specifies a character employed as the first character to denote
a control line to be passed through without processing.
This option applicable to ascii input only.
(# is the default value).
.TP
.BI \-e " string"
.I String
is an arbitrary string to be output if an error is detected during
data transformations.
The default value is: *\et*.
Note that if the
.B \-b,
.B \-i
or
.B \-o
options are employed, an error is returned as HUGE_VAL
value for both return values.
.TP
.BI \-E
causes the input coordinates to be copied to the output line
prior to printing the converted values.
.TP
.BI \-l "[p|P|=|e|u|d]" id
List projection identifiers with
.B \-l,
.B \-lp
or
.B \-lP (expanded)
that can be selected with
.B +proj.
.BI \-l= id
gives expanded description of projection
.I id.
List ellipsoid identifiers with
.B \-le,
that can be selected with
.B +ellps,
.B \-lu
list of cartesian to meter conversion factors
that can be selected with
.B +units
or
.B \-ld
list of datums that can be selected with
.B +datum.
.TP
.BI \-r
This options reverses the order of the
expected input from longitude-latitude or x-y to latitude-longitude or y-x.
.TP
.BI \-s
This options reverses the order of the
output from x-y or longitude-latitude to y-x or latitude-longitude.
.TP
.BI \-S
Causes estimation of
.I meridinal
and
.I parallel
scale factors,
.I area
scale factor and
.I angular distortion,
and
.I maximum
and
.I minimum
scale factors to be listed between <> for each input point.
For conformal projections meridinal and parallel scales factors
will be equal and angular distortion zero.
Equal area projections will have an area factor of 1.
.TP
.BI \-m " mult"
The cartesian data may be scaled by the
.I mult
parameter.
When processing data in a forward projection mode the
cartesian output values are multiplied by
.I mult
otherwise the input cartesian values are divided by
.I mult
before inverse projection.
If the first two characters of
.I mult
are 1/ or 1: then the reciprocal value of
.I mult
is employed.
.TP
.BI \-f " format"
.I Format
is a
.I printf
format string to control the form of the output values.
For inverse projections, the output will be in degrees when this option
is employed.
The default format is "%.2f" for forward projection and DMS
for inverse.
.TP
.BI \-[w|W] n
.I N
is the number of significant fractional digits to employ for
seconds output (when the option is not specified,
.B \-w3
is assumed).
When
.B \-W
is employed the fields will be constant width and with leading zeroes.
.TP
.B \-v
causes a listing of cartographic control parameters tested for and
used by the program to be printed prior to input data.
Should not be used with the
.B \-T
option.
.TP
.B \-V
This option causes an expanded annotated listing of the characteristics
of the projected point.
.B -v is implied with this option.
.TP
.BI \-T " ulow,uhi,vlow,vhi,res[,umax,vmax]"
This option creates a set of bivariate Chebyshev polynomial
coefficients that approximate the selected cartographic projection on
.I stdout.
The values
.I low
and
.I hi
denote the range of the input where the
.I u
or
.I v
prefixes apply to respective longitude-x or latitude-y
depending upon whether a forward or inverse projection is selected.
.I Res
is an integer number specifying the power of 10 precision of the
approximation.
For example, a
.I res
of \-3 specifies an approximation with an accuracy better than .001.
.I Umax,
and
.I vmax
specify maximum degree of the polynomials (default: 15).
See also:
.B fproj(1).
.PP
The
.B +args
run-line arguments are associated with cartographic parameters
and usage varies with projection and for a complete description see
.I "Cartographic Projection Procedures for the UNIX Environment\(emA User's Manual"
and supplementary documentation for Release 4.
.PP
Additional projection control parameters may be contained in two
auxiliary control files:
the first is optionally referenced with the
.BI +init= file:id
and the second is always processed after the name
of the projection has been established from either the run-line
or the contents of
.B +init
file.
The environment parameter
.B PROJ_LIB
establishes the default directory for a file reference without
an absolute path.  This is also used for supporting files like
datum shift files.
.PP
One or more
.I files
(processed in left to right order)
specify the source of data to be transformed.
A \- will specify the location of processing standard input.
If no files are specified, the input is assumed to be from
.I stdin.
For ASCII input data the two data values must be in the
first two white space separated fields and
when both input and output are ASCII all trailing portions
of the input line are appended to the output line.
.PP
Input geographic data
(longitude and latitude) must be in DMS format and input
cartesian data must be in units consistent with the ellipsoid
major axis or sphere radius units.
Output geographic coordinates will be in DMS
(if the
.B \-w
switch is not employed) and precise to 0.001"
with trailing, zero-valued minute-second fields deleted.
.SH EXAMPLE
The following script
.RS 5
 \f(CWproj +proj=utm +lon_0=112w +ellps=clrk66 \-r <<EOF
 45d15'33.1"	111.5W
 45d15.551666667N	\-111d30
 +45.25919444444	111d30'000w
 EOF\fR
.RE
will perform UTM forward projection with a standard UTM
central meridian nearest longitude 112\(deW.
The geographic values of this example are equivalent and meant
as examples of various forms of DMS input.
The x\-y output data will appear as three lines of:
.RS 5
 \f(CW460769.27	5011648.45
.RE
.SH OTHER PROGRAMS
.PP
The \fIproj\fR program is limited to converting between geographic and
projection coordinates within one datum.
.PP
The \fIcs2cs\fR program operates
similarly, but allows translation between any pair of definable coordinate
systems, including support for datum translation.
.PP
The \fIgeod\fR program provides the ability to compute geodesic (Great
Circle) computations.
.SH SEE ALSO
.B cs2cs(1), geod(1), pj_init(3),
.br
.I "Cartographic Projection Procedures for the UNIX Environment\(emA User's Manual,"
(Evenden, 1990, Open-file report 90\-284).
.br
.I "Map Projections Used by the U. S. Geological Survey"
(Snyder, 1984,
USGS Bulletin 1532).
.br
.I "Map Projections\(emA Working Manual"
(Snyder, 1988, USGS Prof. Paper 1395).
.br
.I "An Album of Map Projections"
(Snyder & Voxland, 1989, USGS Prof. Paper 1453).
.SH BUGS
A list of known bugs can found at https://github.com/OSGeo/proj.4/issues
where new bug reports can be submitted too.
.SH HOME PAGE
http://proj4.org/


NEWS
= GDAL/OGR 1.11.3 Release Notes = 

The 1.11.3 release is a bug fix release.

== Build (Unix) ==
 * Support custom buildflags in GRASS plugin (#5850)
 * GRASS 7.0.0 support (#5852)
 * fix detection of ECW SDK 5.1 on Mac (#5867, #5390)
 * Add support for Poppler >= 0.31.0 (#6118)

== Port ==
 * /vsitar/: remove useless validation test that prevents from opening valid .tar files (#5864)

== Algorithms ==
 * gdal_grid: fix crash in optimized mode with GCC 4.4 on 64bit (#5987)
 * Fix failure in GDALSuggestedWarpOut2() when computing output image size in case the top-left and bottom-right corners tranform to the same point (#5980)
 * Rasterize: add check to avoid burning negative x values (#5641)

== GDAL core ==
 * GMLJP2 reader: add compatibility with OGC CRS URL as found in GMLJP2v2 (#5940)
 * GMLJP2: on reading, don't do axis inversation if there's an explicit axisName requesting easting, northing order (#5960)
 * GMLJP2: add missing rangeParameters element to validate against GMLJP2 schema (#5707)

== Utilities ==
 * gdalbuildvrt: fix potential crash when using -b switch (#6095)
 * gdalserver: fix compilation with recent GNU libc (such as in Arch Linux) (#6073)
 * gdalwarp: emit error message if file specified with -cutline cannot be opened (#5921)
 * gdalwarp: fix 1.11.2 regression when invoking several times gdalwarp with several input files on the same target file, and when the input files have a nodata setting: only the last input file was warped onto the target dataset (#5909)
 * ogr2ogr: fix crash with -clipdst when a reprojection fails before (#5973)
 * ogr_layer_algebra.py: for Update, Clip and Erase, only creates attribute of input layer by default (#5976)

== GDAL drivers ==

ECRG driver:
 * change subdataset definition to make sure that they only consist of frames of same scale (#6043)

ERS driver:
 * fix SetProjection() (#5840)

GeoRaster driver:
 * New SRID search (#5911)
 * Fix user-defined SRID issue (#5881)

GTiff driver:
 * fix GTiffDataset::IsBlockAvailable() wrong behaviour when compiling against internal libtiff, when a BigTIFF file has a TileByteCounts with LONG/4-byte counts and not LONG8 (#6001)
 * avoid generated corrupted  right-most and bottom-most tiles for 12-bit JPEG-compressed (#5971)
 * on reading attach PROJ4 extension node for Mercator_Auxiliary_Sphere / WebMercator (#5924)

HDF4 driver:
 * fix AnyTypeToDouble() to use proper type (int instead of long) to work with DFNT_INT32/DFNT_UINT32 on 64-bit Linux (#5965)

HF2 driver:
 * fix reading side of the driver to work on architectures where char is unsigned, like PPC or ARM (#6082)

NetCDF driver:
 * fix crash on opening a NOAA dataset (#5962)
 * fix computation of inverse flattening (#5858)

Northwoord driver:
 * Fix computation of intermediate color table values on non-Intel platforms (#6091)
 * NWT_GRD: don't advertize scale/offset as they are transparently applied in IReadBlock() (#5839).

RPFTOC driver:
 * add tweak for weird relative directory names in the A.TOC file (#5979)

TIL driver:
 * fix half pixel shift in geo registration (#5961)

VRT driver:
 * add more checks to CheckCompatibleForDatasetIO() to avoid issues with overview bands (#5954)
 * fix rounding of output window size on VRTSimpleSource (#5874)

== OGRSpatialReference ==
 * morphFromESRI(): map Mercator_Auxiliary_Sphere to EPSG:3857 (#5924)

== OGR drivers ==

EDIGEO driver:
 * backport conversion of atof() to CPLAtof() to avoid truncation of floating point numbers including coordinates on a non C locale (#6070)

GML driver:
 * accept choice between gml:polygonProperty and gml:multiPolygonProperty (#5977)
 * fix GML_ATTRIBUTES_TO_OGR_FIELDS=YES to work correctly with xlink:href too (#5970); and also backport a trunk fix for writing the correctly path to attributes in nested elements
 * VFR: add support for OriginalniHraniceOmpv (#5853)

GeoJSON driver:
 * avoid truncation of real numbers on reading (#5882)
 * TopoJSON: fix segfault when reading https://raw.githubusercontent.com/mbostock/topojson/master/test/topojson/properties-id-computed.json (#5919)
 * GeoJSON writer: make string comparison for authority name case insensitive so as to recognize lowercase 'epsg' (#4995)

GPKG driver:
 * on reading, recognize GEOMCOLLECTION to be compatible of GDAL 2.0. Still writing GEOMETRYCOLLECTION in 1.11 branch (#5937)
 * escape column names on reading/insert/update (#5879)

MSSQLSpatial driver:
 * Add support for creating features with default values (#5957)

OpenFileGDB driver:
 * fix 1.11.2 regression with FileGDB v9 where the presence of a non spatial table can cause other layers not to be listed (#5875)

PG/PGDump drivers:
 * fix truncation of fields to work with multi-byte UTF-8 characters (#5854)

Shape driver:
 * SHPReadOGRFeature() should free the passed psShape when not NULL and if the DBF record is deleted, so as to avoid later reads to faild (linked to #5974)

SQLite driver:
 * Fix issue with consecutive identical characters in layer name with ExecuteSQL() (#6107)

== SWIG Language Bindings ==

Perl bindings:
 * Fix creation of Field Definition object (#6099)

Python bindings:
 * Fix build with SWIG 3.0.6 (#6045)

= GDAL/OGR 1.11.2 Release Notes = 

The 1.11.2 release is a bug fix release.

== Build ==

  * configure: fix detection of OCI by changing linking order to please modern GCC (#5550)
  * configure: fix test to accept MariaDB 10.X as valid MySQL (#5722)
  * More compiler warnings addressed (#5414)
  * Windows build of PDF: fix compilation issue with Visual Studio 2012 (#5744)
  * Windows build: Add support for MrSID 9.1 SDK (#5814)
  * Windows build: when building netCDF, HDF4, HDF5 as plugins, call registration of 'sub-drivers' GMT, HDF4Image and HDF5Image (#5802)

== Port ==

 * CSV finder: Stop probing for csv/horiz_cs.csv. (#5698)
 * /vsicurl/: avoid reading after end-of-file and fix failure when reading more than 16MB in a single time (#5786)
 * CPLHexToBinary(): faster implementation (#5812)


proj4/src/proj.c
/* <<<< Cartographic projection filter program >>>> */
#include "projects.h"
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <math.h>
#include "emess.h"

/* TK 1999-02-13 */
#if defined(MSDOS) || defined(OS2) || defined(WIN32) || defined(__WIN32__)
#  include <fcntl.h>
#  include <io.h>
#  define SET_BINARY_MODE(file) setmode(fileno(file), O_BINARY)
#else
#  define SET_BINARY_MODE(file)
#endif
/* ! TK 1999-02-13 */

#define MAX_LINE 1000
#define MAX_PARGS 100
#define PJ_INVERS(P) (P->inv ? 1 : 0)
	static PJ
*Proj;
	static projUV
(*proj)(projUV, PJ *);
	static int
reversein = 0,	/* != 0 reverse input arguments */
reverseout = 0,	/* != 0 reverse output arguments */
bin_in = 0,	/* != 0 then binary input */
bin_out = 0,	/* != 0 then binary output */
echoin = 0,	/* echo input data to output line */
tag = '#',	/* beginning of line tag character */
inverse = 0,	/* != 0 then inverse projection */
prescale = 0,	/* != 0 apply cartesian scale factor */
dofactors = 0,	/* determine scale factors */
facs_bad = 0,	/* return condition from pj_factors */
very_verby = 0, /* very verbose mode */
postscale = 0;
	static char
*cheby_str,		/* string controlling Chebychev evaluation */
*oform = (char *)0,	/* output format for x-y or decimal degrees */
*oterr = "*\t*",	/* output line for unprojectable input */
*usage =
"%s\nusage: %s [ -bCeEfiIlormsStTvVwW [args] ] [ +opts[=arg] ] [ files ]\n";
	static struct FACTORS
facs;
	static double
(*informat)(const char *, char **),	/* input data deformatter function */
fscale = 0.;	/* cartesian scale factor */
	static projUV
int_proj(projUV data) {
	if (prescale) { data.u *= fscale; data.v *= fscale; }
	data = (*proj)(data, Proj);
	if (postscale && data.u != HUGE_VAL)
		{ data.u *= fscale; data.v *= fscale; }
	return(data);
}
	static void	/* file processing function */
process(FILE *fid) {
	char line[MAX_LINE+3], *s, pline[40];
	projUV data;

	for (;;) {
		++emess_dat.File_line;
		if (bin_in) {	/* binary input */
			if (fread(&data, sizeof(projUV), 1, fid) != 1)
				break;
		} else {	/* ascii input */
			if (!(s = fgets(line, MAX_LINE, fid)))
				break;
			if (!strchr(s, '\n')) { /* overlong line */
				int c;
				(void)strcat(s, "\n");
				/* gobble up to newline */
				while ((c = fgetc(fid)) != EOF && c != '\n') ;
			}
			if (*s == tag) {
				if (!bin_out)
					(void)fputs(line, stdout);
				continue;
			}
			if (reversein) {
				data.v = (*informat)(s, &s);
				data.u = (*informat)(s, &s);
			} else {
				data.u = (*informat)(s, &s);
				data.v = (*informat)(s, &s);
			}
			if (data.v == HUGE_VAL)
				data.u = HUGE_VAL;
			if (!*s && (s > line)) --s; /* assumed we gobbled \n */
			if (!bin_out && echoin) {
				int t;
				t = *s;
				*s = '\0';
				(void)fputs(line, stdout);
				*s = t;
				putchar('\t');
			}
		}
		if (data.u != HUGE_VAL) {
			if (prescale) { data.u *= fscale; data.v *= fscale; }
			if (dofactors && !inverse)
				facs_bad = pj_factors(data, Proj, 0., &facs);
			data = (*proj)(data, Proj);
			if (dofactors && inverse)
				facs_bad = pj_factors(data, Proj, 0., &facs);
			if (postscale && data.u != HUGE_VAL)
				{ data.u *= fscale; data.v *= fscale; }
		}
		if (bin_out) { /* binary output */
			(void)fwrite(&data, sizeof(projUV), 1, stdout);
			continue;
		} else if (data.u == HUGE_VAL) /* error output */
			(void)fputs(oterr, stdout);
		else if (inverse && !oform) {	/*ascii DMS output */
			if (reverseout) {
				(void)fputs(rtodms(pline, data.v, 'N', 'S'), stdout);
				putchar('\t');
				(void)fputs(rtodms(pline, data.u, 'E', 'W'), stdout);
			} else {
				(void)fputs(rtodms(pline, data.u, 'E', 'W'), stdout);
				putchar('\t');
				(void)fputs(rtodms(pline, data.v, 'N', 'S'), stdout);
			}
		} else {	/* x-y or decimal degree ascii output */
			if (inverse) {
				data.v *= RAD_TO_DEG;
				data.u *= RAD_TO_DEG;
			}
			if (reverseout) {
				(void)printf(oform,data.v); putchar('\t');
				(void)printf(oform,data.u);
			} else {
				(void)printf(oform,data.u); putchar('\t');
				(void)printf(oform,data.v);
			}
		}
		if (dofactors) /* print scale factor data */
                {
			if (!facs_bad)
				(void)printf("\t<%g %g %g %g %g %g>",
					facs.h, facs.k, facs.s,
					facs.omega * RAD_TO_DEG, facs.a, facs.b);
			else
				(void)fputs("\t<* * * * * *>", stdout);
                }
		(void)fputs(bin_in ? "\n" : s, stdout);
	}
}
	static void	/* file processing function --- verbosely */
vprocess(FILE *fid) {
	char line[MAX_LINE+3], *s, pline[40];
	projUV dat_ll, dat_xy;
	int linvers;

	if (!oform)
		oform = "%.3f";
	if (bin_in || bin_out)
		emess(1,"binary I/O not available in -V option");
	for (;;) {
		++emess_dat.File_line;
		if (!(s = fgets(line, MAX_LINE, fid)))
			break;
		if (!strchr(s, '\n')) { /* overlong line */
			int c;
			(void)strcat(s, "\n");
			/* gobble up to newline */
			while ((c = fgetc(fid)) != EOF && c != '\n') ;
		}
		if (*s == tag) { /* pass on data */
			(void)fputs(s, stdout);
			continue;
		}
		/* check to override default input mode */
		if (*s == 'I' || *s == 'i') {
			linvers = 1;
			++s;
		} else if (*s == 'I' || *s == 'i') {
			linvers = 0;
			++s;
		} else
			linvers = inverse;
		if (linvers) {
			if (!PJ_INVERS(Proj)) {
				emess(-1,"inverse for this projection not avail.\n");
				continue;
			}
			dat_xy.u = strtod(s, &s);
			dat_xy.v = strtod(s, &s);
			if (dat_xy.u == HUGE_VAL || dat_xy.v == HUGE_VAL) {
				emess(-1,"lon-lat input conversion failure\n");
				continue;
			}
			if (prescale) { dat_xy.u *= fscale; dat_xy.v *= fscale; }
			dat_ll = pj_inv(dat_xy, Proj);
		} else {
			dat_ll.u = dmstor(s, &s);
			dat_ll.v = dmstor(s, &s);
			if (dat_ll.u == HUGE_VAL || dat_ll.v == HUGE_VAL) {
				emess(-1,"lon-lat input conversion failure\n");
				continue;
			}
			dat_xy = pj_fwd(dat_ll, Proj);
			if (postscale) { dat_xy.u *= fscale; dat_xy.v *= fscale; }
		}
		if (pj_errno) {
			emess(-1, pj_strerrno(pj_errno));
			continue;
		}
		if (!*s && (s > line)) --s; /* assumed we gobbled \n */
		if (pj_factors(dat_ll, Proj, 0., &facs)) {
			emess(-1,"failed to conpute factors\n\n");
			continue;
		}
		if (*s != '\n')
			(void)fputs(s, stdout);
		(void)fputs("Longitude: ", stdout);
		(void)fputs(rtodms(pline, dat_ll.u, 'E', 'W'), stdout);
		(void)printf(" [ %.11g ]\n", dat_ll.u * RAD_TO_DEG);
		(void)fputs("Latitude:  ", stdout);
		(void)fputs(rtodms(pline, dat_ll.v, 'N', 'S'), stdout);
		(void)printf(" [ %.11g ]\n", dat_ll.v * RAD_TO_DEG);
		(void)fputs("Easting (x):   ", stdout);
		(void)printf(oform, dat_xy.u); putchar('\n');
		(void)fputs("Northing (y):  ", stdout);
		(void)printf(oform, dat_xy.v); putchar('\n');
		(void)printf("Meridian scale (h)%c: %.8f  ( %.4g %% error )\n",
			facs.code & IS_ANAL_HK ? '*' : ' ', facs.h, (facs.h-1.)*100.);
		(void)printf("Parallel scale (k)%c: %.8f  ( %.4g %% error )\n",
			facs.code & IS_ANAL_HK ? '*' : ' ', facs.k, (facs.k-1.)*100.);
		(void)printf("Areal scale (s):     %.8f  ( %.4g %% error )\n",
			facs.s, (facs.s-1.)*100.);
		(void)printf("Angular distortion (w): %.3f\n", facs.omega *
			RAD_TO_DEG);
		(void)printf("Meridian/Parallel angle: %.5f\n",
			facs.thetap * RAD_TO_DEG);
		(void)printf("Convergence%c: ",facs.code & IS_ANAL_CONV ? '*' : ' ');
		(void)fputs(rtodms(pline, facs.conv, 0, 0), stdout);
		(void)printf(" [ %.8f ]\n", facs.conv * RAD_TO_DEG);
		(void)printf("Max-min (Tissot axis a-b) scale error: %.5f %.5f\n\n",
			facs.a, facs.b);
	}
}

int main(int argc, char **argv) {
    char *arg, **eargv = argv, *pargv[MAX_PARGS], **iargv = argv;
    FILE *fid;
    int pargc = 0, iargc = argc, eargc = 0, c, mon = 0;

    if ( (emess_dat.Prog_name = strrchr(*argv,DIR_CHAR)) != NULL)
        ++emess_dat.Prog_name;
    else emess_dat.Prog_name = *argv;
    inverse = ! strncmp(emess_dat.Prog_name, "inv", 3);
    if (argc <= 1 ) {
        (void)fprintf(stderr, usage, pj_get_release(), emess_dat.Prog_name);
        exit (0);
    }
    /* process run line arguments */
    while (--argc > 0) { /* collect run line arguments */
        if(**++argv == '-') for(arg = *argv;;) {
            switch(*++arg) {
              case '\0': /* position of "stdin" */
                if (arg[-1] == '-') eargv[eargc++] = "-";
                break;
              case 'b': /* binary I/O */
                bin_in = bin_out = 1;
                continue;
              case 'C': /* Check - run internal regression tests */
                return pj_run_selftests (very_verby);
                continue;
              case 'v': /* monitor dump of initialization */
                mon = 1;
                continue;
              case 'i': /* input binary */
                bin_in = 1;
                continue;
              case 'o': /* output binary */
                bin_out = 1;
                continue;
              case 'I': /* alt. method to spec inverse */
                inverse = 1;
                continue;
              case 'E': /* echo ascii input to ascii output */
                echoin = 1;
                continue;
              case 'V': /* very verbose processing mode */
                very_verby = 1;
                mon = 1;
              case 'S': /* compute scale factors */
                dofactors = 1;
                continue;
              case 't': /* set col. one char */
                if (arg[1]) tag = *++arg;
                else emess(1,"missing -t col. 1 tag");
                continue;
              case 'l': /* list projections, ellipses or units */
                if (!arg[1] || arg[1] == 'p' || arg[1] == 'P') {
                    /* list projections */
                    struct PJ_LIST *lp;
                    int do_long = arg[1] == 'P', c;
                    char *str;

                    for (lp = pj_get_list_ref() ; lp->id ; ++lp) {
                        if( strcmp(lp->id,"latlong") == 0
                            || strcmp(lp->id,"longlat") == 0
                            || strcmp(lp->id,"geocent") == 0 )
                            continue;

                        (void)printf("%s : ", lp->id);
                        if (do_long)  /* possibly multiline description */
                            (void)puts(*lp->descr);
                        else { /* first line, only */
                            str = *lp->descr;
                            while ((c = *str++) && c != '\n')
                                putchar(c);
                            putchar('\n');
                        }
                    }
                } else if (arg[1] == '=') { /* list projection 'descr' */
                    struct PJ_LIST *lp;

                    arg += 2;
                    for (lp = pj_get_list_ref(); lp->id ; ++lp)
                        if (!strcmp(lp->id, arg)) {
                            (void)printf("%9s : %s\n", lp->id, *lp->descr);
                            break;
                        }
                } else if (arg[1] == 'e') { /* list ellipses */
                    struct PJ_ELLPS *le;

                    for (le = pj_get_ellps_ref(); le->id ; ++le)
                        (void)printf("%9s %-16s %-16s %s\n",
                                     le->id, le->major, le->ell, le->name);
                } else if (arg[1] == 'u') { /* list units */
                    struct PJ_UNITS *lu;

                    for (lu = pj_get_units_ref(); lu->id ; ++lu)
                        (void)printf("%12s %-20s %s\n",
                                     lu->id, lu->to_meter, lu->name);
                } else if (arg[1] == 'd') { /* list datums */
                    struct PJ_DATUMS *ld;

                    printf("__datum_id__ __ellipse___ __definition/comments______________________________\n" );
                    for (ld = pj_get_datums_ref(); ld->id ; ++ld)
                    {
                        printf("%12s %-12s %-30s\n",
                               ld->id, ld->ellipse_id, ld->defn);
                        if( ld->comments != NULL && strlen(ld->comments) > 0 )
                            printf( "%25s %s\n", " ", ld->comments );
                    }
                } else
                    emess(1,"invalid list option: l%c",arg[1]);
                exit(0);
                continue; /* artificial */
              case 'e': /* error line alternative */
                if (--argc <= 0)
                    noargument:
                emess(1,"missing argument for -%c",*arg);
                oterr = *++argv;
                continue;
              case 'T': /* generate Chebyshev coefficients */
                if (--argc <= 0) goto noargument;
                cheby_str = *++argv;
                continue;
              case 'm': /* cartesian multiplier */
                if (--argc <= 0) goto noargument;
                postscale = 1;
                if (!strncmp("1/",*++argv,2) ||
                    !strncmp("1:",*argv,2)) {
                    if((fscale = atof((*argv)+2)) == 0.)
                        goto badscale;
                    fscale = 1. / fscale;
                } else
                    if ((fscale = atof(*argv)) == 0.) {
                      badscale:
                        emess(1,"invalid scale argument");
                    }
                continue;
              case 'W': /* specify seconds precision */
              case 'w': /* -W for constant field width */
                if ((c = arg[1]) != 0 && isdigit(c)) {
                    set_rtodms(c - '0', *arg == 'W');
                    ++arg;
                } else
                    emess(1,"-W argument missing or non-digit");
                continue;
              case 'f': /* alternate output format degrees or xy */
                if (--argc <= 0) goto noargument;
                oform = *++argv;
                continue;
              case 'r': /* reverse input */
                reversein = 1;
                continue;
              case 's': /* reverse output */
                reverseout = 1;
                continue;
              default:
                emess(1, "invalid option: -%c",*arg);
                break;
            }
            break;
        } else if (**argv == '+') { /* + argument */
            if (pargc < MAX_PARGS)
                pargv[pargc++] = *argv + 1;
            else
                emess(1,"overflowed + argument table");
        } else /* assumed to be input file name(s) */
            eargv[eargc++] = *argv;
    }
    if (eargc == 0 && !cheby_str) /* if no specific files force sysin */
        eargv[eargc++] = "-";
    else if (eargc > 0 && cheby_str) /* warning */
        emess(4, "data files when generating Chebychev prohibited");
    /* done with parameter and control input */
    if (inverse && postscale) {
        prescale = 1;
        postscale = 0;
        fscale = 1./fscale;
    }
    if (!(Proj = pj_init(pargc, pargv)))
        emess(3,"projection initialization failure\ncause: %s",
              pj_strerrno(pj_errno));

    if( pj_is_latlong( Proj ) )
    {
        emess( 3, "+proj=latlong unsuitable for use with proj program." );
        exit( 0 );
    }

    if (inverse) {
        if (!Proj->inv)
            emess(3,"inverse projection not available");
        proj = pj_inv;
    } else
        proj = pj_fwd;
    if (cheby_str) {
        extern void gen_cheb(int, projUV(*)(projUV), char *, PJ *, int, char **);

        gen_cheb(inverse, int_proj, cheby_str, Proj, iargc, iargv);
        exit(0);
    }
    /* set input formating control */
    if (mon) {
        pj_pr_list(Proj);
        if (very_verby) {
            (void)printf("#Final Earth figure: ");
            if (Proj->es) {
                (void)printf("ellipsoid\n#  Major axis (a): ");
                (void)printf(oform ? oform : "%.3f", Proj->a);
                (void)printf("\n#  1/flattening: %.6f\n",
                             1./(1. - sqrt(1. - Proj->es)));
                (void)printf("#  squared eccentricity: %.12f\n", Proj->es);
            } else {
                (void)printf("sphere\n#  Radius: ");
                (void)printf(oform ? oform : "%.3f", Proj->a);
                (void)putchar('\n');
            }
        }
    }
    if (inverse)
        informat = strtod;
    else {
        informat = dmstor;
        if (!oform)
            oform = "%.2f";
    }

    if (bin_out)
    {
        SET_BINARY_MODE(stdout);
    }

    /* process input file list */
    for ( ; eargc-- ; ++eargv) {
        if (**eargv == '-') {
            fid = stdin;
            emess_dat.File_name = "<stdin>";

            if (bin_in)
            {
                SET_BINARY_MODE(stdin);
            }

        } else {
            if ((fid = fopen(*eargv, "rb")) == NULL) {
                emess(-2, *eargv, "input file");
                continue;
            }
            emess_dat.File_name = *eargv;
        }
        emess_dat.File_line = 0;
        if (very_verby)
            vprocess(fid);
        else
            process(fid);
        (void)fclose(fid);
        emess_dat.File_name = 0;
    }
    if( Proj )
        pj_free(Proj);
    exit(0); /* normal completion */
}
